File: python/algorithms/louvain.cpp
#include "ukv/ukv.hpp"

using namespace unum::ukv;
using namespace unum;

struct community_degree_t {
    double in_degree {};
    double tot_degree {};
};

using partition_t = std::unordered_map<ukv_key_t, ukv_key_t>;
using vertex_degrees_t = std::unordered_map<ukv_key_t, double>;
using graph_t = std::unordered_map<ukv_key_t, std::unordered_map<ukv_key_t, double>>;
using community_degrees_t = std::unordered_map<ukv_key_t, community_degree_t>;

bool first_phase(graph_collection_t& graph,
                 partition_t& partition,
                 vertex_degrees_t& degrees,
                 community_degrees_t& community_degrees,
                 std::size_t count_edges) {

    bool improvement = false;
    bool modified = true;
    auto stream = graph.vertex_stream().throw_or_release();
    vertex_degrees_t degree_in_coms;

    while (modified) {
        modified = false;
        stream.seek_to_first();
        while (!stream.is_end()) {
            auto vertex = stream.key();
            auto vertex_degree = degrees[vertex];
            auto vertex_community = partition[vertex];
            auto best_mod = 0.0;
            auto best_com = vertex_community;
            auto vertex_com_tot_degree = community_degrees[vertex_community].tot_degree;

            degree_in_coms.clear();
            auto neighbors = graph.neighbors(vertex).throw_or_release();
            for (auto neighbor : neighbors)
                degree_in_coms[partition[neighbor]] += 1;
            auto vertex_in_vertex_com_degree = degree_in_coms[vertex_community];

            for (auto neighbor : neighbors) {
                auto neighbor_community = partition[neighbor];
                if (vertex_community == neighbor_community)
                    continue;

                auto neighbor_com_tot_degree = community_degrees[neighbor_community].tot_degree;
                auto vertex_in_neighbor_com_degree = degree_in_coms[neighbor_community];

                double delta = (1.0 / count_edges) * (vertex_in_neighbor_com_degree - vertex_in_vertex_com_degree) -
                               (vertex_degree / (2.0 * (count_edges * count_edges))) *
                                   (vertex_degree + neighbor_com_tot_degree - vertex_com_tot_degree);

                if (delta > best_mod) {
                    best_mod = delta;
                    best_com = neighbor_community;
                }
            }

            if (best_com != vertex_community) {
                community_degrees[vertex_community].tot_degree -= (vertex_degree - vertex_in_vertex_com_degree);
                auto vertex_in_best_com_degree = degree_in_coms[best_com];
                community_degrees[best_com].tot_degree += (vertex_degree - vertex_in_best_com_degree);
                community_degrees[vertex_community].in_degree -= vertex_in_vertex_com_degree;
                community_degrees[best_com].in_degree += vertex_in_best_com_degree;
                partition[vertex] = best_com;
                modified = true;
                improvement = true;
            }
            ++stream;
        }
    }

    return improvement;
}

bool second_phase(graph_t& graph,
                  partition_t& partition,
                  vertex_degrees_t& degrees,
                  community_degrees_t& community_degrees,
                  std::size_t count_edges) {

    bool improvement = false;
    bool modified = true;
    vertex_degrees_t degree_in_coms;

    while (modified) {
        modified = false;
        for (auto const& graph_iter : graph) {
            auto const& vertex = graph_iter.first;
            auto const& neighbors = graph_iter.second;
            auto vertex_degree = degrees[vertex];
            auto vertex_community = partition[vertex];
            auto best_mod = 0.0;
            auto best_com = vertex_community;
            auto vertex_com_tot_degree = community_degrees[vertex_community].tot_degree;

            degree_in_coms.clear();
            for (auto const& neighbor_iter : neighbors)
                degree_in_coms[partition[neighbor_iter.first]] += neighbor_iter.second;
            auto vertex_in_vertex_com_degree = degree_in_coms[vertex_community];

            for (auto const& neighbor_iter : neighbors) {
                auto neighbor_community = partition[neighbor_iter.first];
                if (vertex_community == neighbor_community)
                    continue;

                auto neighbor_com_tot_degree = community_degrees[neighbor_community].tot_degree;
                auto vertex_in_neighbor_com_degree = degree_in_coms[neighbor_community];

                double delta = (1.0 / count_edges) * (vertex_in_neighbor_com_degree - vertex_in_vertex_com_degree) -
                               (vertex_degree / (2.0 * (count_edges * count_edges))) *
                                   (vertex_degree + neighbor_com_tot_degree - vertex_com_tot_degree);

                if (delta > best_mod) {
                    best_mod = delta;
                    best_com = neighbor_community;
                }
            }

            if (best_com != vertex_community) {
                community_degrees[vertex_community].tot_degree -= (vertex_degree - vertex_in_vertex_com_degree);
                auto vertex_in_best_com_degree = degree_in_coms[best_com];
                community_degrees[best_com].tot_degree += (vertex_degree - vertex_in_best_com_degree);
                community_degrees[vertex_community].in_degree -= vertex_in_vertex_com_degree;
                community_degrees[best_com].in_degree += vertex_in_best_com_degree;
                partition[vertex] = best_com;
                modified = true;
                improvement = true;
            }
        }
    }

    return improvement;
}

double modularity(partition_t& partition, community_degrees_t& community_degrees, double deg_sum) {
    auto m = deg_sum / 2;
    auto norm = 1 / (m * m);
    double res = 0.0;
    for (auto const& iter : partition)
        res += ((community_degrees[iter.second].tot_degree - community_degrees[iter.second].in_degree) / m) -
               (deg_sum * deg_sum * norm);
    return res;
}

graph_t gen_graph(graph_collection_t& graph, partition_t& partition) {

    graph_t induced_graph;
    auto stream = graph.vertex_stream().throw_or_release();

    while (!stream.is_end()) {
        auto vertex = stream.key();
        ++stream;

        auto com1 = partition[vertex];
        auto neighbors = graph.neighbors(vertex).throw_or_release();
        for (auto neighbor : neighbors) {
            auto com2 = partition[neighbor];
            if (com1 == com2)
                continue;
            induced_graph[com1][com2] += 1;
        }
    }

    return induced_graph;
}

graph_t gen_graph(graph_t& graph, partition_t& partition) {

    graph_t induced_graph;
    for (auto const& graph_iter : graph) {
        auto com1 = partition[graph_iter.first];
        for (auto const& neighbor_iter : graph_iter.second) {
            auto com2 = partition[neighbor_iter.first];
            if (com1 == com2)
                continue;
            induced_graph[com1][com2] += neighbor_iter.second;
        }
    }

    return induced_graph;
}

partition_t best_partition(graph_collection_t& graph_collection, float min_modularity_growth = 0.0000001) {

    auto count_vertices = graph_collection.number_of_vertices();
    auto count_edges = graph_collection.number_of_edges();

    partition_t partition;
    std::vector<partition_t> partitions;
    vertex_degrees_t vertex_degrees;
    community_degrees_t community_degrees;

    partition.reserve(count_vertices);
    vertex_degrees.reserve(count_vertices);
    community_degrees.reserve(count_vertices);

    auto stream = graph_collection.vertex_stream().throw_or_release();
    while (!stream.is_end()) {
        auto vertices = stream.keys_batch();
        auto degrees = graph_collection.degrees(vertices).throw_or_release();
        for (std::size_t i = 0; i != vertices.size(); ++i) {
            partition[vertices[i]] = vertices[i];
            vertex_degrees[vertices[i]] = degrees[i];
            community_degrees[vertices[i]] = {0, degrees[i]};
        }
        stream.seek_to_next_batch();
    }

    bool improvement = first_phase(graph_collection, partition, vertex_degrees, community_degrees, count_edges);
    auto mod = modularity(partition, community_degrees, count_edges);
    auto graph = gen_graph(graph_collection, partition);
    partitions.insert(partitions.begin(), std::move(partition));

    while (improvement) {
        partition_t partition;
        partition.reserve(graph.size());
        vertex_degrees.clear();
        community_degrees.clear();

        auto degree_sum = 0;
        count_edges = 0;
        for (auto const& graph_iter : graph) {
            partition[graph_iter.first] = graph_iter.first;
            double degree = 0;
            for (auto const& neighbor_iter : graph_iter.second) {
                degree += neighbor_iter.second;
                count_edges += 1;
            }
            degree_sum += degree;
            vertex_degrees[graph_iter.first] = degree;
            community_degrees[graph_iter.first] = {0, degree};
        }

        improvement = second_phase(graph, partition, vertex_degrees, community_degrees, count_edges / 2);
        auto new_mod = modularity(partition, community_degrees, degree_sum / 2);
        if (new_mod - mod <= min_modularity_growth)
            break;

        graph = gen_graph(graph, partition);
        partitions.insert(partitions.begin(), std::move(partition));
        mod = new_mod;
    }

    partition = partitions[0];
    for (std::size_t index = 1; index != partitions.size(); ++index) {
        for (auto const& iter : partitions[index])
            partitions[index][iter.first] = partition[iter.second];
        partition = partitions[index];
    }

    return partition;
}

File: python/algorithms/louvain.py
# Baseline Python implementation of Louvain Community detection algorithm,
# similar to the one in NetworkX, designed to operate with both in-memory
# graphs and the ones persisted in UStore.
from typing import Mapping, Tuple

import networkx as nx


def _node_affiliation(graph, partition: Mapping[int, int], node: int) -> Tuple[int, int]:
    """
    Simultaneously computes the degree (without self-edges) between `node` and
    members of `node_community` as well as `neighbor_community`.
    """

    degrees = {}
    for neighbor in graph[node]:
        if neighbor != node:
            degrees[partition[neighbor]] = degrees.get(
                partition[neighbor], 0) + 1
    return degrees


def _node_affiliation_weighted(graph, partition: Mapping[int, int], node: int) -> Tuple[int, int]:
    """
    Simultaneously computes the weighted degree (without self-edges) between `node` and
    members of `node_community` (without counting self-edge) as well as `neighbor_community`.
    """

    degrees = {}
    for neighbor, data in graph[node].items():
        if neighbor != node:
            degrees[partition[neighbor]] = degrees.get(
                partition[neighbor], 0) + data.get('weight', 1)
    return degrees


def _one_level(graph, edges_weight_sum: float, is_weighted: bool) \
    -> tuple[
        Mapping[int, int],
        Mapping[int, float],
        Mapping[int, float],
        bool
]:
    """
    Compute one level communities of graph.
    """

    partition = {}
    degrees = {}
    # For every partition, stores the accumulated number of degrees of all of its members
    community_degrees = {}
    # For every partition, stores the accumulated number of degrees between all of its members
    community_in_degrees = {}

    if is_weighted:
        for node in graph.nodes:
            community_in_degrees[node] = 0
            degrees[node] = graph.degree(weight='weight')[node]
            partition[node] = node
    else:
        for node in graph.nodes:
            community_in_degrees[node] = 0
            degrees[node] = graph.degree[node]
            partition[node] = node

    community_degrees = degrees.copy()
    affiliation_func = _node_affiliation_weighted if is_weighted else _node_affiliation

    improvement = False
    modified = 1
    while modified:
        modified = 0
        for node in graph.nodes:
            node_degree = degrees[node]
            node_community = partition[node]
            best_mod = 0
            best_com = node_community
            node_com_tot_degree = community_degrees[node_community]

            degree_in_coms = affiliation_func(graph, partition, node)
            node_in_node_com_degree = degree_in_coms.get(node_community, 0)

            for neighbor in graph[node]:

                neighbor_community = partition[neighbor]
                if node_community == neighbor_community:
                    continue

                neighbor_com_tot_degree = community_degrees[neighbor_community]
                node_in_neighbor_com_degree = degree_in_coms[neighbor_community]

                delta_modularity = (1 / edges_weight_sum) * (node_in_neighbor_com_degree - node_in_node_com_degree) + \
                    (node_degree / (2 * (edges_weight_sum ** 2))) * \
                    (node_com_tot_degree - node_degree - neighbor_com_tot_degree)

                if delta_modularity > best_mod:
                    best_mod = delta_modularity
                    best_com = neighbor_community

            if best_com != node_community:
                community_degrees[node_community] -= node_degree
                node_in_best_com_degree = degree_in_coms[best_com]
                community_degrees[best_com] += node_degree
                community_in_degrees[node_community] -= node_in_node_com_degree
                community_in_degrees[best_com] += node_in_best_com_degree
                partition[node] = best_com
                modified += 1
                improvement = True

            # TODO: remove this awkward logic to assume NetworkX Graph here
            # But still we need to add weights for self-edges in community_in
            if is_weighted:
                edge_weight = graph.get_edge_data(node, node)
                community_in_degrees[best_com] += edge_weight.get(
                    'weight', 1) if edge_weight is not None else 0
            else:
                community_in_degrees[best_com] += 1 if graph.has_edge(
                    node, node) else 0

    return partition, community_degrees, community_in_degrees, improvement


def _gen_graph(graph, partition: Mapping[int, int]):
    """
    Create super node graph, where nodes are communities and edges are degrees between communities.
    """

    G = nx.Graph()
    G.add_nodes_from(partition.values())

    for node1, node2 in graph.edges:
        com1 = partition[node1]
        com2 = partition[node2]
        w_prec = G.get_edge_data(com1, com2, {'weight': 0}).get('weight', 1)
        G.add_edge(com1, com2, **{'weight': w_prec + 1})

    return G


def _gen_graph_from_weighted(graph, partition: Mapping[int, int]):
    """
    Create super node graph from weighted graph, where nodes are communities and edges are degrees between communities.
    """

    G = nx.Graph()
    G.add_nodes_from(partition.values())

    for node1, node2, data in graph.edges(data=True):
        com1 = partition[node1]
        com2 = partition[node2]
        edge_weight = data.get('weight', 1)
        w_prec = G.get_edge_data(com1, com2, {'weight': 0}).get('weight', 1)
        G.add_edge(com1, com2, **{'weight': w_prec + edge_weight})

    return G


def modularity(
        partition: Mapping[int, int],
        community_degrees: Mapping[int, float],
        community_in_degrees: Mapping[int, float],
        degree_sum: float) -> float:
    """
    Compute the modularity of a partition of a graph
    """

    m = degree_sum / 2
    norm = 1 / m
    res = 0.
    for com in set(partition.values()):
        res += norm * (community_in_degrees[com] -
                       community_degrees[com] * community_degrees[com] / (2*m))
    return res


def best_partition(graph, min_mod_growth: float = 0.0000001) -> Mapping[int, int]:
    """
    Compute the partition of the graph nodes which maximises the modularity.
    """

    partitions = []
    count_edges = graph.number_of_edges()
    partition, community_degrees, community_in_degrees, improvement = _one_level(
        graph, count_edges, False)
    if not improvement:
        return partition
    mod = modularity(partition, community_degrees,
                     community_in_degrees, 2 * count_edges)

    partitions.insert(0, partition)
    graph = _gen_graph(graph, partition)

    while improvement:
        edges_weight_sum = graph.size(weight='weight')
        partition, community_degrees, community_in_degrees, improvement = _one_level(
            graph, edges_weight_sum, True)
        new_mod = modularity(partition, community_degrees,
                             community_in_degrees, 2 * edges_weight_sum)

        if not improvement or new_mod - mod <= min_mod_growth:
            break

        mod = new_mod
        partitions.insert(0, partition)
        graph = _gen_graph_from_weighted(graph, partition)

    partition = partitions[0]
    for index in range(1, len(partitions)):
        for node, com in partitions[index].items():
            partitions[index][node] = partition[com]
        partition = partitions[index]

    return partition
